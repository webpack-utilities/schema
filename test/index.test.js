const { validateOptions } = require('../src/index.js')

describe('Error', () => {
  test('Success', () => {
    const options = {
      type() {},
      array: [ '' ],
      string: '',
      object: {
        prop: false,
        object: {
          prop: false,
        },
      },
      boolean: true,
      instance: new RegExp('')
    }

    expect(
      validateOptions('test/fixtures/schema.json', options, 'Loader')
    ).toBe(true)
  })

  test('Throws', () => {
    const options = {
      type: null,
      array: {},
      string: false,
      object: {
        prop: 1,
        object: false,
      },
      boolean: '',
      instance() {}
    }

    const validate = () => {
      return validateOptions('test/fixtures/schema.json', options, '{Name}')
    }

    expect(validate).toThrowError()
    expect(validate).toThrowErrorMatchingSnapshot()
  })

  test('Errors', () => {
    const options = {
      type: null,
      array: {},
      string: false,
      object: {
        prop: 1,
        object: false,
      },
      boolean: '',
      instance() {}
    }

    const validate = () => {
      return validateOptions('test/fixtures/schema.json', options, '{Name}')
    }

    try {
      validate()
    } catch (err) {
      const errors = err.errors.map((err) => err.dataPath)

      const expected = [
        '/string',
        '/array',
        '/object/prop',
        '/object/object',
        '/boolean',
        '/type',
        '/instance'
      ]

      expect(errors).toMatchObject(expected)
      expect(err.errors).toMatchSnapshot()
    }
  })

  describe('Messages', () => {
    test('Default', () => {
      const options = {
        type() {},
        array: [ '' ],
        string: 1,
        object: {
          prop: false,
          object: {
            prop: false,
          },
        },
        boolean: true,
        instance: new RegExp('')
      }

      const validate = () => {
        return validateOptions(
          'test/fixtures/schema.json',
          options,
          '{Name}'
        )
      }

      try {
        validate()
      } catch (err) {
        err.errors.forEach((err) => expect(err).toMatchSnapshot())

        expect(err.message).toMatchSnapshot()
      }
    })

    test('Custom', () => {
      const options = {
        type() {},
        array: [ '' ],
        string: 1,
        object: {
          prop: false,
          object: {
            prop: false,
          },
        },
        boolean: true,
        instance: new RegExp('')
      }

      const validate = () => {
        return validateOptions(
          'test/fixtures/errors/schema.json',
          options,
          '{Name}'
        )
      }

      try {
        validate()
      } catch (err) {
        err.errors.forEach((err) => expect(err).toMatchSnapshot())

        expect(err.message).toMatchSnapshot()
      }
    })
  })
})
