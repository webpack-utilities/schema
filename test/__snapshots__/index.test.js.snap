// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error Errors 1`] = `
Array [
  Object {
    "dataPath": "/string",
    "keyword": "type",
    "message": "should be string",
    "params": Object {
      "type": "string",
    },
    "schemaPath": "#/properties/string/type",
  },
  Object {
    "dataPath": "/array",
    "keyword": "type",
    "message": "should be array",
    "params": Object {
      "type": "array",
    },
    "schemaPath": "#/properties/array/type",
  },
  Object {
    "dataPath": "/object/prop",
    "keyword": "type",
    "message": "should be boolean",
    "params": Object {
      "type": "boolean",
    },
    "schemaPath": "#/properties/object/properties/prop/type",
  },
  Object {
    "dataPath": "/object/object",
    "keyword": "type",
    "message": "should be object",
    "params": Object {
      "type": "object",
    },
    "schemaPath": "#/properties/object/properties/object/type",
  },
  Object {
    "dataPath": "/boolean",
    "keyword": "type",
    "message": "should be boolean",
    "params": Object {
      "type": "boolean",
    },
    "schemaPath": "#/properties/boolean/type",
  },
  Object {
    "dataPath": "/type",
    "keyword": "typeof",
    "message": "should pass \\"typeof\\" keyword validation",
    "params": Object {
      "keyword": "typeof",
    },
    "schemaPath": "#/properties/type/typeof",
  },
  Object {
    "dataPath": "/instance",
    "keyword": "instanceof",
    "message": "should pass \\"instanceof\\" keyword validation",
    "params": Object {
      "keyword": "instanceof",
    },
    "schemaPath": "#/properties/instance/instanceof",
  },
]
`;

exports[`Error Messages Custom 1`] = `
Object {
  "dataPath": "/string",
  "keyword": "errorMessage",
  "message": "should be {String} (https://github.com/org/repo#anchor)",
  "params": Object {
    "errors": Array [
      Object {
        "dataPath": "/string",
        "keyword": "type",
        "message": "should be string",
        "params": Object {
          "type": "string",
        },
        "schemaPath": "#/properties/string/type",
      },
    ],
  },
  "schemaPath": "#/errorMessage",
}
`;

exports[`Error Messages Custom 2`] = `
"{Name} Invalid Options

options.string should be {String} (https://github.com/org/repo#anchor)
"
`;

exports[`Error Messages Default 1`] = `
Object {
  "dataPath": "/string",
  "keyword": "type",
  "message": "should be string",
  "params": Object {
    "type": "string",
  },
  "schemaPath": "#/properties/string/type",
}
`;

exports[`Error Messages Default 2`] = `
"{Name} Invalid Options

options.string should be string
"
`;

exports[`Error Throws 1`] = `
"{Name} Invalid Options

options.string should be string
options.array should be array
options.object.prop should be boolean
options.object.object should be object
options.boolean should be boolean
options.type should pass \\"typeof\\" keyword validation
options.instance should pass \\"instanceof\\" keyword validation
"
`;
